uber/config.py

```diff
@@ -13,15 +13,9 @@ def mixin(cls, klass):
     def include_plugin_config(self, plugin_config):
         """Plugins call this method to merge their own config into the global c object."""
 
-        for attr, val in plugin_config.items():
-            if not isinstance(val, dict):
-                setattr(self, attr.upper(), val)
-
-        if 'enums' in plugin_config:
-            self.make_enums(plugin_config['enums'])
-
-        if 'dates' in plugin_config:
-            self.make_dates(plugin_config['dates'])
```

Comment:
> I don't think those other things are actually getting overridden in the c object

The way you've implemented it here, some of them will be and some will not.  For example:
- ``c.VOLUNTEER_CHECKLIST`` gets parsed lower in ``config.py`` so ``c.config_base`` will have updated values based on the plugin but ``c.VOLUNTEER_CHECKLIST`` will remain unchanged.
- ``c.DEPT_HEAD_CHECKLIST`` is just set to the config section, so plugins are able to affect both the values in ``c.config_base`` and ``c.DEPT_HEAD_CHECKLIST`` simply by including a ``[dept_head_checklist]`` section in their own config

That's probably undesirable, and these kinds of edge cases were why I originally kept ``c.include_plugin_config`` limited to just top-level items, dates, and enums.  If we want to do a full ``c.config_base.merge`` then I think we should take the effort to make some guarantees that plugin config will successfully be able to set/override all values.

> Hm, on second thought, I think that may be leaving out top-level single config items.

Yes, this current implementation does leave that out, so that will need to be added back in.


However, there's a larger (and subtler) problem.  Suppose that I define a plugin with its own ``[enums]`` section with a ``[[job_location]]`` subsection which adds a single new job location, e.g. ``hotel = 'Hotel Workers'`` or something like that.  This will cause the following sequence of things to happen at import time:
- ``c.make_enums`` is called initially here in ``config.py`` which creates (among other things) ``c.JOB_LOCATION_OPTS``; this is a list which does NOT have the ``hotel`` option
- ``uber.models`` is imported, and the ``Attendee`` class is defined with its ``assigned_depts`` column, which references ``c.JOB_LOCATION_OPTS`` (without the hotel option)
- the plugin is imported, which calls ``c.include_plugin_config``
- this causes a whole new ``c.JOB_LOCATION_OPTS`` list to be created which DOES have the new hotel option, but the old list is thrown away

When all is said and done, the old list which was previously assigned to ``c.JOB_LOCATION_OPTS`` was discarded for the new list, but it had already been referenced.  This is why we need to be updating these lists and dictionaries in place.  Theoretically we could instead refactor all of the places in our codebase which reference them at import time and just literally never reference anything in ``c`` at import time, though I'd rather just update in place and make sure that all of our code is written with the assumption that these updates-in-place can occur.


In case you can't tell, this level of complexity is something I punted on when originally implementing ``c.include_plugin_config``, but it would definitely be nice to have eventually.  I figured that I'd just work on several plugins and see what use cases turned out to be useful in practice and let that drive the implementation instead of tackling a complicated refactor in advance.  (Not that I thought through everything that would be involved, like I didn't really considered much of this at the time, I just generally shied away from adding/refactoring stuff in the core plugin unless/until it became obvious that it was necessary.)

So with all of this being said, what are your thoughts on moving forward?  Do you want to try to work through the different challenges of config merging now, or hold off on this feature until we've either got time to really dive into it and/or it becomes really useful/necessary/timesaving?  (I confess I don't remember offhand why this came up in the first place so maybe it's already necessary and I just forgot.)